### PR TITLE
lib2geom: init at 1.0

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change
https://inkscape.org/news/2018/11/11/graphics-math-library-2geoms-first-release-availab/

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by jtojnar** [comment]
```
20: Test command: /home/jtojnar/projects/nixpkgs/lib2geom-1.0.0/build/src/tests/implicitization-test
20: Test timeout computed to be: 10000000
20: /home/jtojnar/projects/nixpkgs/lib2geom-1.0.0/build/src/tests/implicitization-test: error while loading shared libraries: libgtest.so: cannot open shared object file: No such file or directory
20/20 Test #20: implicitization-test .............***Failed    0.00 sec
```

```ShellSession
$ ldd lib2geom-1.0.0/build/src/tests/implicitization-test
	linux-vdso.so.1 (0x00007ffcd334b000)
	[…]
	libgtest.so => not found
```

**Comment by jtojnar** [comment]
Hmm, it does not use system gtest at all: https://bugs.launchpad.net/lib2geom/+bug/1803860

File: pkgs/development/libraries/lib2geom/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  # TODO: move this into ninja setup-hook
+  checkPhase = ''
+    ninja test
+  '';
+
+  meta = with stdenv.lib; { 
+    description = "Easy to use 2D geometry library in C++";
+    homepage = https://gitlab.com/inkscape/lib2geom;
+    license = [ licenses.lgpl21Plus licenses.mpl11 ];
```
Comment (c0bw3b) [inline pkgs/development/libraries/lib2geom/default.nix]:
```suggestion
    license = [ licenses.lgpl21 licenses.mpl11 ];
```
Headers in source files don't seem to mention "or any later version" for the LGPL

Comment (jtojnar) [reply pkgs/development/libraries/lib2geom/default.nix]: Thank you.

**Comment by jtojnar** [comment]
The tests failing are just CMake not using RPATH for build by default. After fixing that, I encountered some [test failure](https://gitlab.com/inkscape/lib2geom/issues/5).

**Comment by jtojnar** [comment]
I pushed the alpha version of Inkscape so we could test this better.

**Comment by jtojnar** [comment]
Main issue with this is that the icons are not loaded correctly for some reason. Maybe because they have been moved but I would expect that to cause issues on other platforms as well.

**Comment by jtojnar** [comment]
Opened an upstream issue about the icons: https://gitlab.com/inkscape/inkscape/issues/135

**Comment by jtojnar** [comment]
I updated to alpha2 but since Inkscape recently [improved their CMake files](https://gitlab.com/inkscape/inkscape/commit/8294023227f1ec5d71956ee5e6cd46a4f03b2b93#9a2aa4db38d3115ed60da621e012c0efc0172aae_59_60), the `RPATH` is broken.

It contains `$ORIGIN/..//nix/store/7gifw2g1qkjvqf7bdkhxxhxzy49980zs-inkscape-1.0-alpha2/lib/inkscape` instead of the previous `$ORIGIN/../lib/inkscape`, or our expected `/nix/store/7gifw2g1qkjvqf7bdkhxxhxzy49980zs-inkscape-1.0-alpha2/lib/inkscape`.

I am trying to figure out if CMake supports something as trivial as Meson’s `join_path` function. At the moment, I am drawing blanks:

https://stackoverflow.com/questions/57382992/joining-paths-in-cmake?

**Comment by flosse** [comment]
@jtojnar Thanks for your PR :)

I just tried to install 18420530975cc77345bc0d94ce5d53a8a0369528 on NixOS 19.03 (`cd nixpkgs/ && nix-env -iA inkscape1 -f .`) with the following result:
```sh
inkscape --version
/home/mk/.nix-profile/bin/inkscape: error while loading shared libraries: libinkscape_base.so: cannot open shared object file: No such file or directory
```

**Comment by jtojnar** [comment]
Hmm, RPATH contains `$ORIGIN/..//nix/store/jm7jm40j116jvn7gvrv4as930was154q-inkscape-1.0-beta1/lib/inkscape`.

**Comment by jtojnar** [comment]
Yeah, see above https://github.com/NixOS/nixpkgs/pull/50286#issuecomment-518845346

**Comment by jtojnar** [comment]
beta2 fixed the issue in https://gitlab.com/inkscape/inkscape/commit/c77532d0acf37d0e1d9ebc46f1bc0a05ffca01e8

**Comment by prusnak** [comment]
I haven't noticed this PR when creating mine, but I think https://github.com/NixOS/nixpkgs/pull/85169 supersedes this one. Let me tell you why:

* extensions are now included as a submodule (`share/extensions`), no need to fetch them separately 
* same for lib2geom which is now kept as an in-tree copy (`src/2geom`)

The only thing I am unsure about is `glib-networking` which is present in your changeset, but I am unsure why you've added it. Can you please elaborate?

**Comment by worldofpeace** [comment]
> The only thing I am unsure about is glib-networking which is present in your changeset, but I am unsure why you've added it. Can you please elaborate?

It's probably for the glib-networking gio module to be added to the wrapper, which is needed for https stuff I believe.

**Comment by prusnak** [comment]
OK, I will amend the other PR to include `glib-networking`.

**Comment by jtojnar** [comment]
lib2geom has always been part of the source tree. The main change of inkscape 1.0 is that we can finally use lib2geom from Nixpkgs and we should do so as vendored dependencies are :vomiting_face:

**Comment by jtojnar** [comment]
The reason I am fetching the extensions separately is that `fetchFromGitLab` does not support submodules :crying_cat_face:

**Comment by jtojnar** [comment]
Hmm, apparently they still do not allow non-vendored lib2geom:

https://gitlab.com/inkscape/inkscape/-/merge_requests/1410

**Comment by jtojnar** [comment]
Let's just use this PR for lib2geom introduction as I originally intended.

**Review by worldofpeace — APPROVED** [review]
`lib2geom` expression lgtm.

**Comment by worldofpeace** [comment]
oh, there's a darwin problem
```
[11/95] Building CXX object tests/CMakeFiles/rotated-rect-test.dir/rotated-rect-test.cpp.o
FAILED: tests/CMakeFiles/rotated-rect-test.dir/rotated-rect-test.cpp.o
/nix/store/1ngwf6wvh9hf1vz3h2h5znqi90fa86q1-clang-wrapper-7.1.0/bin/clang++   -I/nix/store/m3j6p2gzazmh3fx39kz1l01xmifaqbqn-glib-2.64.1-dev/include/glib-2.0 -I/nix/store/vhp0hj6r62cj7y1wbwy885xxr1f2m42a-glib-2.64.1/lib/glib-2.0/include -I../include -I../include/2geom -arch x86_64   -std=c++11 -MD -MT tests/CMakeFiles/rotated-rect-test.dir/rotated-rect-test.cpp.o -MF tests/CMakeFiles/rotated-rect-test.dir/rotated-rect-test.cpp.o.d -o tests/CMakeFiles/rotated-rect-test.dir/rotated-rect-test.cpp.o -c ../tests/rotated-rect-test.cpp
In file included from ../tests/rotated-rect-test.cpp:35:
../include/2geom/rotated-rect.h:79:32: error: implicit instantiation of undefined template 'std::__1::array<Geom::Point, 4>'
    const std::array<Point, 4> m_corners;
                               ^
/nix/store/47m4h1bb5df8k6031qbljfr4w3m63f68-libc++-7.1.0/include/c++/v1/__tuple:223:64: note: template is declared here
template <class _Tp, size_t _Size> struct _LIBCPP_TEMPLATE_VIS array;
                                                               ^
../tests/rotated-rect-test.cpp:117:1: warning: 'InstantiateTestCase_P_IsDeprecated' is deprecated: INSTANTIATE_TEST_CASE_P is deprecated, please use INSTANTIATE_TEST_SUITE_P [-Wdeprecated-declarations]
INSTANTIATE_TEST_CASE_P(intesect_non_aligned, RotatedRectTest, testing::Values(
^
/nix/store/hjysj1qari3harv79iv4qn777hwvx54n-gtest-1.10.0-dev/include/gtest/gtest-param-test.h:496:38: note: expanded from macro 'INSTANTIATE_TEST_CASE_P'
  static_assert(::testing::internal::InstantiateTestCase_P_IsDeprecated(), \
                                     ^
/nix/store/hjysj1qari3harv79iv4qn777hwvx54n-gtest-1.10.0-dev/include/gtest/internal/gtest-internal.h:1207:1: note: 'InstantiateTestCase_P_IsDeprecated' has been explicitly marked deprecated here
GTEST_INTERNAL_DEPRECATED(
^
/nix/store/hjysj1qari3harv79iv4qn777hwvx54n-gtest-1.10.0-dev/include/gtest/internal/gtest-port.h:2224:59: note: expanded from macro 'GTEST_INTERNAL_DEPRECATED'
#define GTEST_INTERNAL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                          ^
1 warning and 1 error generated.
```

**Comment by prusnak** [comment]
Out of curiosity, Jan, are there any other projects (other than Inkscape which uses it as an in-tree copy) that use this library?

**Comment by jtojnar** [comment]
I am not aware of any but it is still good idea to unvendor it. We do not have as strong anti-vendoring policies as Fedora or Debian but using dependencies from Nix is still good here, if just for making the builds more parallelizable.

**Comment by jtojnar** [comment]
Though of course it does not make much sense to merge it without any consumers.

**Comment by jtojnar** [comment]
Added in https://github.com/NixOS/nixpkgs/pull/124333

**Pull request closed without merging**